CONTRIBUTING.md

```diff
@@ -43,18 +43,102 @@ Commits should also be self-contained, meaning that they should pass all tests a
 
 ### Commit messages
 
-We use [conventional commits](https://www.conventionalcommits.org/) for our commit messages.
+Commit messages should be written in the imperative mood.
+For example, use "Add feature" instead of "Added feature".
+Also, they should be written in impersonal form, meaning that you should not use terms like "I" or "we".
+
+This project requires that commit messages for the [conventional commits](https://www.conventionalcommits.org/) specification.
 This means that each commit message should follow the format `<type>(<scope>): <description>`.
+The description should start with a capital letter and not end with a period.
+
+#### Types
+
 The valid types are defined in the [Commitlint configuration file](.commitlintrc.yml).
+When chosing a type, please consider the following guidelines:
+
+* `build`: Changes that affect the build system.
+* `chore`: Changes that do not affect the application, e.g., removing an obsolete workaround in the code.
+* `ci`: Changes to the CI configuration.
+* `deps`: Changes to the dependencies of the project.
+* `docs`: Changes to the documentation, either in code or in documentation files.
+* `feat`: A new feature of the application.
+* `fix`: A bug fix in the application.
+* `perf`: A change that improves the performance of the application.
+* `refactor`: A code refactoring that does not change the behavior of the application.
+* `revert`: A commit that reverts a previous commit.
+* `style`: Changes that do not affect the meaning of the code, e.g., formatting changes.
+* `test`: Changes to the tests of the application.
+
+The types `fix`, `feat`, and `perf` are reserved for changes that affect the application and should not be used for changes that are not user-facing.
+For example, a new build system feature should use `build`, and a test fix should use `test`.
+This is important because the commit message titles are used to generate the release notes, and the feature and fix sections in the release notes should only contain user-facing changes.
+
+#### Scopes
+
 The scopes should represent the part of the project that was changed, e.g., `analyzer`, `core`, or `ui`.
+This usually matches the name of the Gradle module that was changed.
 If a commit touches multiple parts of the project, and it is not possible to split it into multiple commits, the scope can be omitted.
 
-Commit messages should be written in the imperative mood.
-For example, use "Add feature" instead of "Added feature".
+#### Breaking changes
+
+Breaking changes should be indicated by either adding a `BREAKING CHANGE:` section to the commit message body or by adding a `!` after the type, for example:
```

Comment:
We probably should communicate what is considered a breaking change, so is the REST API only thing we consider for versioning. Also, if we plan to do API versioning (`/api/v1/` etc.) what exactly is a breaking change if we strictly follow that?

Reply:
@mmurto I have created a separate issue for this: #1172